### DISCUSSION TITLE
[2023-05-29] GUI Team Demo Day

### DISCUSSION THREAD

**Discussion by wdanilo** [enso-org/enso#6876] [Weekly Demo Day]
Hi! :) 

From my end, I was working on the new FRP implementation. This week I was mostly working on its performance and the way we can implement everything in the most performant way. The current state is:
- The new implementation can initialize FRP nodes 10-15x faster than the old one.
- Sending data via the new FRP nodes is as fast as in the old implementation if the data is copyable.
- Sending data via the new FRP nodes is 2-10x faster than in the old implementation if the data is more complex and requires clones on every FRP node in the old implementation (the new one sends references to data whenever possible).
- The new implementation uses very tricky structures with plenty of unsafe code that I refactored into well documented libraries with plenty of tests to keep it safe.
- Due to several health issues I was working a little bit slower last week than I anticipated, and thus I will have probably a few days delay on this task (it was planned for Tuesday). I will try to finish it till the end of the week.
- What is left is adding missing FRP nodes, adding tests, cleaning the code and a LOT of testing if it works instead of the old implementation.

**Comment by MichaelMauderer** [comment]
* Fixed project initialisation performance regression #6758 

* Worked on fixing multiple bugs caused by our overly complex visualization FRP logic in #6831 

* Empty Visualization when opening a full-screen visualization directly without opening the visualization before. #6770

https://github.com/enso-org/enso/assets/1428930/5812ed03-652c-4a27-8e33-b85512ca11b6

* Empty visualization when opening the full-screen visualization before the data for the visualization has arrived. #6561

https://github.com/enso-org/enso/assets/1428930/d8e58f2d-f1b6-4b70-84fa-e917f6c0af1f

* Visualization is reset to default when reconnecting nodes #6673

https://github.com/enso-org/enso/assets/1428930/ac6cf79a-7147-4f13-9045-52599fb39900


* Redundant internal open/lose events caused by logic loops around the show/hide button, as well as many redundant layer setting/unsetting issues internal to the visualization code. 

* Fixed visualizations being closed when evens arrive in an unexpected (but valid) order.  #6766

**Comment by mwu-tow** [comment]
Last week summary:
1) Developing provisional logging infrastructure to ease debugging (redirecting renderer logs to the file; diffing module changes);
1) Created PR with several "safe" URM improvements (https://github.com/enso-org/enso/pull/6877);
1) Developed a workaround for issues with drag&drop node creation (https://github.com/enso-org/enso/issues/6591) but decided against applying it right now this due to concerns about the cloud upload scenario performance; 
1) Ongoing work on fixing behavior of some "not-as-atomic-as-should" actions (like removing connected node);
1) Debugging race conditions and issues that can happen when dealing with asynchronous updates.

**Comment by farmaazon** [comment]
### The CB improvements were polished

Also added some fixes, e.g. for unexpected spaces inside literals. As a rule, you should no longer be forced to use `ctrl+enter` if inserting actually existing stuff from CB.

https://github.com/enso-org/enso/assets/3919101/dc3e6aeb-3b22-4997-bf4b-3e70a03ce9fc

Still not landed, as the QA get prolonged.

### Fixed parts of CB input being cut off when accepting too early.

https://github.com/enso-org/enso/assets/3919101/5c508004-91d3-4523-bb10-af9be3fa0c3f

Also, fixed dashboard button/status bar displayed over full-screen visualization.

Combined with a significant refactoring of our FRP in the Project View.

**Comment by Frizi** [comment]
- Finished fixes related to invalid widget selection for various expressions https://github.com/enso-org/enso/issues/6760
- Implemented nested widget definitions (dynamic widgets for parameters inside a dropdown entry). Finished automatic import resolution and generation of short labels for both static and dynamic dropdowns. https://github.com/enso-org/enso/pull/6878

<img width="772" alt="image" src="https://github.com/enso-org/enso/assets/919491/97c70654-9170-4cf0-ae4d-2c25c74caa96">

<img width="247" alt="Pasted image 20230526181540" src="https://github.com/enso-org/enso/assets/919491/27e2077e-27ca-4d22-8479-e0d3d54a1674">

- Took part in expanding expression info data to include more data about method application https://github.com/orgs/enso-org/discussions/6762
- Created a discussion about missing features needed to support more complicated dynamic widgets (e.g. proper dropdown values for `Table.join` column widget). https://github.com/orgs/enso-org/discussions/6832
- Fixed some additional minor issues related to widgets (e.g. https://github.com/enso-org/enso/issues/6825)

**Comment by kazcw** [comment]
I've been ensuring the new edges in #6707 match the layout of the old edges. It is a matter of porting the old layout logic (which is often radius-first) to logic better suited to the new monoshape implementation (which is (Δx, Δy) based). To check that the algorithms produce the same results, I've added the old edges to the debug scene; they will only be seen if there are differences. 

[vokoscreenNG-2023-05-29_07-31-55.webm](https://github.com/enso-org/enso/assets/1047859/72dc8c49-5bdc-4463-94c5-5b12c55b8bba)

There's one remaining known difference (seen at the end of the video). I'll be fixing that case this morning.
